File: UniversalDeviceToolkit.WPF/Flags.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using LenovoLegionToolkit.Lib.Utils;

// ReSharper disable StringLiteralTypo

namespace UniversalDeviceToolkit.WPF;

public class Flags
{
    public const string DisableUpdateCheckerSwitch = "--disable-update-checker";

    public bool IsTraceEnabled { get; }
    public bool Minimized { get; }
    public bool DisableTrayTooltip { get; }
    public bool AllowAllPowerModesOnBattery { get; }
    public bool ForceDisableRgbKeyboardSupport { get; }
    public bool ForceDisableSpectrumKeyboardSupport { get; }
    public bool ForceDisableLenovoLighting { get; }
    public bool ExperimentalGPUWorkingMode { get; }
    public Uri? ProxyUrl { get; }
    public string? ProxyUsername { get; }
    public string? ProxyPassword { get; }
    public bool ProxyAllowAllCerts { get; }
    public bool DisableUpdateChecker { get; }

    public Flags(IEnumerable<string> startupArgs)
    {
        var args = startupArgs.Concat(LoadExternalArgs()).ToArray();

        IsTraceEnabled = BoolValue(args, "--trace");
        Minimized = BoolValue(args, "--minimized");
        DisableTrayTooltip = BoolValue(args, "--disable-tray-tooltip");
        AllowAllPowerModesOnBattery = BoolValue(args, "--allow-all-power-modes-on-battery");
        ForceDisableRgbKeyboardSupport = BoolValue(args, "--force-disable-rgbkb");
        ForceDisableSpectrumKeyboardSupport = BoolValue(args, "--force-disable-spectrumkb");
        ForceDisableLenovoLighting = BoolValue(args, "--force-disable-lenovolighting");
        ExperimentalGPUWorkingMode = BoolValue(args, "--experimental-gpu-working-mode");
        ProxyUrl = Uri.TryCreate(StringValue(args, "--proxy-url"), UriKind.Absolute, out var uri) ? uri : null;
        ProxyUsername = StringValue(args, "--proxy-username");
        ProxyPassword = StringValue(args, "--proxy-password");
        ProxyAllowAllCerts = BoolValue(args, "--proxy-allow-all-certs");
        DisableUpdateChecker = BoolValue(args, DisableUpdateCheckerSwitch);
    }

    private static string[] LoadExternalArgs()
    {
        try
        {
            var argsFile = Path.Combine(Folders.AppData, "args.txt");
            return !File.Exists(argsFile) ? [] : File.ReadAllLines(argsFile);
        }
        catch
        {
            return [];
        }
    }

    private static bool BoolValue(IEnumerable<string> values, string key) => values.Contains(key);

    internal static string? StringValue(IEnumerable<string> values, string key)
    {
        var args = values.ToArray();

        for (var i = 0; i < args.Length; i++)
        {
            var value = args[i];
            if (value.Equals(key, StringComparison.OrdinalIgnoreCase))
                return i + 1 < args.Length && !args[i + 1].StartsWith("--", StringComparison.Ordinal)
                    ? args[i + 1]
                    : null;

            if (value.StartsWith($"{key}=", StringComparison.OrdinalIgnoreCase))
                return value[(key.Length + 1)..];
        }

        return null;
    }

    public override string ToString() =>
        $"{nameof(IsTraceEnabled)}: {IsTraceEnabled}," +
        $" {nameof(Minimized)}: {Minimized}," +
        $" {nameof(DisableTrayTooltip)}: {DisableTrayTooltip}," +
        $" {nameof(AllowAllPowerModesOnBattery)}: {AllowAllPowerModesOnBattery}," +
        $" {nameof(ForceDisableRgbKeyboardSupport)}: {ForceDisableRgbKeyboardSupport}," +
        $" {nameof(ForceDisableSpectrumKeyboardSupport)}: {ForceDisableSpectrumKeyboardSupport}," +
        $" {nameof(ForceDisableLenovoLighting)}: {ForceDisableLenovoLighting}," +
        $" {nameof(ExperimentalGPUWorkingMode)}: {ExperimentalGPUWorkingMode}," +
        $" {nameof(ProxyUrl)}: {ProxyUrl}," +
        $" {nameof(ProxyUsername)}: {ProxyUsername}," +
        $" {nameof(ProxyPassword)}: [REDACTED]," +
        $" {nameof(ProxyAllowAllCerts)}: {ProxyAllowAllCerts}," +
        $" {nameof(DisableUpdateChecker)}: {DisableUpdateChecker}";
}

File: UniversalDeviceToolkit.Electron/src/main/flags.ts
import { app } from 'electron'
import { existsSync, readFileSync } from 'fs'
import { join } from 'path'

/**
 * Command-line flags for the Electron client. Mirrors the Electron app's
 * UniversalDeviceToolkit.Electron.Flags surface: same switch names, same
 * semantics, and the same external args.txt source so launch arguments
 * behave identically between the two clients.
 *
 * Switches that affect the headless host (trace / safe-start / proxy) are
 * forwarded via {@link toHostArgs}; UI-owned switches are consumed here.
 */
export interface AppFlags {
  isTraceEnabled: boolean
  minimized: boolean
  disableTrayTooltip: boolean
  allowAllPowerModesOnBattery: boolean
  forceDisableRgbKeyboardSupport: boolean
  forceDisableSpectrumKeyboardSupport: boolean
  forceDisableLenovoLighting: boolean
  experimentalGpuWorkingMode: boolean
  /** Chromium single-process mode: merge renderers into one process for memory inspection. */
  singleProcess: boolean
  proxyUrl?: string
  proxyUsername?: string
  proxyPassword?: string
  proxyAllowAllCerts: boolean
  disableUpdateChecker: boolean
  safeStart: boolean
  resetHardwareState: boolean
  resetNetworkState: boolean
  restoreProcessorMinState: boolean
  /** Host-only switches: never re-apply hardware state / load plugins. */
  noPlugins: boolean
  noHardware: boolean
}

const BOOL_SWITCHES: ReadonlyArray<readonly [string, keyof AppFlags]> = [
  ['--trace', 'isTraceEnabled'],
  ['--minimized', 'minimized'],
  ['--disable-tray-tooltip', 'disableTrayTooltip'],
  ['--allow-all-power-modes-on-battery', 'allowAllPowerModesOnBattery'],
  ['--force-disable-rgbkb', 'forceDisableRgbKeyboardSupport'],
  ['--force-disable-spectrumkb', 'forceDisableSpectrumKeyboardSupport'],
  ['--force-disable-lenovolighting', 'forceDisableLenovoLighting'],
  ['--experimental-gpu-working-mode', 'experimentalGpuWorkingMode'],
  ['--single-process', 'singleProcess'],
  ['--proxy-allow-all-certs', 'proxyAllowAllCerts'],
  ['--disable-update-checker', 'disableUpdateChecker'],
  ['--safe-start', 'safeStart'],
  ['--reset-hardware-state', 'resetHardwareState'],
  ['--reset-network-state', 'resetNetworkState'],
  ['--restore-processor-min-state', 'restoreProcessorMinState'],
  ['--no-plugins', 'noPlugins'],
  ['--no-hardware', 'noHardware']
]

const VALUE_SWITCHES: ReadonlyArray<readonly [string, keyof AppFlags]> = [
  ['--proxy-url', 'proxyUrl'],
  ['--proxy-username', 'proxyUsername'],
  ['--proxy-password', 'proxyPassword']
]

/**
 * Mirrors Electron Flags.StringValue: matches the key case-insensitively, taking
 * the following argument unless it starts with "--", and also accepts the
 * "--key=value" form.
 */
function stringValue(args: string[], key: string): string | undefined {
  for (let i = 0; i < args.length; i++) {
    const value = args[i]
    if (value.toLowerCase() === key) {
      const next = args[i + 1]
      return next !== undefined && !next.startsWith('--') ? next : undefined
    }
    if (value.toLowerCase().startsWith(`${key}=`)) {
      return value.slice(key.length + 1)
    }
  }
  return undefined
}

export function parseFlags(argv: string[]): AppFlags {
  const flags: AppFlags = {
    isTraceEnabled: false,
    minimized: false,
    disableTrayTooltip: false,
    allowAllPowerModesOnBattery: false,
    forceDisableRgbKeyboardSupport: false,
    forceDisableSpectrumKeyboardSupport: false,
    forceDisableLenovoLighting: false,
    experimentalGpuWorkingMode: false,
    singleProcess: false,
    proxyAllowAllCerts: false,
    disableUpdateChecker: false,
    safeStart: false,
    resetHardwareState: false,
    resetNetworkState: false,
    restoreProcessorMinState: false,
    noPlugins: false,
    noHardware: false
  }

  const record = flags as unknown as Record<string, unknown>

  for (const [switchName, key] of BOOL_SWITCHES) {
    if (argv.includes(switchName)) {
      record[key] = true
    }
  }

  for (const [switchName, key] of VALUE_SWITCHES) {
    const value = stringValue(argv, switchName)
    if (value !== undefined) {
      record[key] = value
    }
  }

  return flags
}

/**
 * Mirrors Electron Flags.LoadExternalArgs: reads extra switches from
 * %LOCALAPPDATA%\UniversalDeviceToolkit\args.txt (one switch per line),
 * matching the Lib Folders.AppData root. Missing/unreadable files are ignored.
 */
export function loadExternalArgs(): string[] {
  try {
    const localAppData = process.env.LOCALAPPDATA
    const root = localAppData
      ? join(localAppData, 'UniversalDeviceToolkit')
      : app.getPath('userData')
    const argsFile = join(root, 'args.txt')
    if (!existsSync(argsFile)) return []
    return readFileSync(argsFile, 'utf8')
      .split(/\r?\n/)
      .map((line) => line.trim())
      .filter((line) => line.length > 0)
  } catch {
    return []
  }
}

/** App-wide flags, parsed once from process.argv plus the external args.txt. */
export const flags: AppFlags = parseFlags([...process.argv.slice(2), ...loadExternalArgs()])

/**
 * Subset of switches understood by the headless host (HostFlags). UI-owned
 * switches (--minimized, --disable-tray-tooltip, ...) are consumed here and
 * deliberately not forwarded.
 */
export function toHostArgs(appFlags: AppFlags): string[] {
  const args: string[] = []
  if (appFlags.isTraceEnabled) args.push('--trace')
  if (appFlags.safeStart) args.push('--safe-start')
  if (appFlags.noPlugins) args.push('--no-plugins')
  if (appFlags.noHardware) args.push('--no-hardware')
  if (appFlags.proxyUrl) args.push('--proxy-url', appFlags.proxyUrl)
  if (appFlags.proxyUsername) args.push('--proxy-username', appFlags.proxyUsername)
  if (appFlags.proxyPassword) args.push('--proxy-password', appFlags.proxyPassword)
  if (appFlags.proxyAllowAllCerts) args.push('--proxy-allow-all-certs')
  return args
}

/** Mirrors Electron Flags.ToString() for trace logging. */
export function describeFlags(appFlags: AppFlags): string {
  return [
    `isTraceEnabled: ${appFlags.isTraceEnabled}`,
    `minimized: ${appFlags.minimized}`,
    `disableTrayTooltip: ${appFlags.disableTrayTooltip}`,
    `allowAllPowerModesOnBattery: ${appFlags.allowAllPowerModesOnBattery}`,
    `forceDisableRgbKeyboardSupport: ${appFlags.forceDisableRgbKeyboardSupport}`,
    `forceDisableSpectrumKeyboardSupport: ${appFlags.forceDisableSpectrumKeyboardSupport}`,
    `forceDisableLenovoLighting: ${appFlags.forceDisableLenovoLighting}`,
    `experimentalGpuWorkingMode: ${appFlags.experimentalGpuWorkingMode}`,
    `singleProcess: ${appFlags.singleProcess}`,
    `proxyUrl: ${appFlags.proxyUrl ?? 'null'}`,
    `proxyUsername: ${appFlags.proxyUsername ?? 'null'}`,
    'proxyPassword: [REDACTED]',
    `proxyAllowAllCerts: ${appFlags.proxyAllowAllCerts}`,
    `disableUpdateChecker: ${appFlags.disableUpdateChecker}`,
    `safeStart: ${appFlags.safeStart}`,
    `resetHardwareState: ${appFlags.resetHardwareState}`,
    `resetNetworkState: ${appFlags.resetNetworkState}`,
    `restoreProcessorMinState: ${appFlags.restoreProcessorMinState}`
  ].join(', ')
}
